From 535cadac594a9c1ae7b5f2c41cf2a536573dbe2a Mon Sep 17 00:00:00 2001
From: Raj Kumar Bhagat <quic_rajkbhag@quicinc.com>
Date: Tue, 18 Jul 2023 18:36:24 +0530
Subject: [PATCH 2/2] ath11k: apply real noise floor offset instead of
 ATH11K_DEFAULT_NOISE_FLOOR

ATH11K_DEFAULT_NOISE_FLOOR is the approximate noise floor value of
the HW, that is used for rssi calculation. Hence use real noise floor
offset value for the calculation of the signal strength. This offset
is saved in ath11k structure (ar->rssi_offsets.rssi_offset) and it is
updated periodically by firmware through wmi event
WMI_PDEV_RSSI_DBM_CONVERSION_PARAMS_INFO_EVENTID.

For the Firmware that do not send this wmi event, the backward
compatibility is maintained by using the existing method to calculate
rssi in dbm if these new offset is not present.

Signed-off-by: Raj Kumar Bhagat <quic_rajkbhag@quicinc.com>
---
 drivers/net/wireless/ath/ath11k/dp_rx.c | 5 +++--
 drivers/net/wireless/ath/ath11k/dp_tx.c | 9 +++++----
 drivers/net/wireless/ath/ath11k/mac.c   | 7 ++++---
 drivers/net/wireless/ath/ath11k/wmi.c   | 3 ++-
 4 files changed, 14 insertions(+), 10 deletions(-)

--- a/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -6213,7 +6213,9 @@ static void ath11k_update_radiotap(struc
 	u16 ampdu_id = ppduinfo->ampdu_id[ppduinfo->userid];
 
 	rxs->flag |= RX_FLAG_MACTIME_START;
-	rxs->signal = ppduinfo->rssi_comb + ATH11K_DEFAULT_NOISE_FLOOR;
+	rxs->signal = ar->rssi_offsets.rssi_offset ?
+		      ppduinfo->rssi_comb + ar->rssi_offsets.rssi_offset :
+		      ppduinfo->rssi_comb + ATH11K_DEFAULT_NOISE_FLOOR;
 
 	if (ppduinfo->nss)
 		rxs->nss = ppduinfo->nss;
@@ -6598,7 +6600,10 @@ int ath11k_dp_rx_process_mon_status(stru
 			if (peer && !peer->sta) {
 				list_for_each_entry_safe(nrp, tmp, &ab->neighbor_peers, list) {
 					if (nrp->is_filter_on && ether_addr_equal(nrp->addr, peer->addr)) {
-						nrp->rssi = ppdu_info->rssi_comb;
+						nrp->rssi = ar->rssi_offsets.rssi_offset ?
+							    ppdu_info->rssi_comb +
+							    ar->rssi_offsets.rssi_offset :
+							    ppdu_info->rssi_comb;
 						nrp->timestamp = ktime_to_ms(ktime_get_real());
 						complete(&nrp->filter_done);
 					}
--- a/drivers/net/wireless/ath/ath11k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_tx.c
@@ -581,8 +581,9 @@ ath11k_dp_tx_htt_tx_complete_buf(struct
 		if (!(info->flags & IEEE80211_TX_CTL_NO_ACK) &&
 		    !(flags & ATH11K_SKB_F_NOACK_TID)) {
 			info->flags |= IEEE80211_TX_STAT_ACK;
-			info->status.ack_signal = ATH11K_DEFAULT_NOISE_FLOOR +
-						  ts->ack_rssi;
+			info->status.ack_signal = ar->rssi_offsets.rssi_offset ?
+						  ts->ack_rssi + ar->rssi_offsets.rssi_offset :
+						  ts->ack_rssi + ATH11K_DEFAULT_NOISE_FLOOR;
 			info->status.flags |=
 				IEEE80211_TX_STATUS_ACK_SIGNAL_VALID;
 		} else {
@@ -919,7 +920,9 @@ static void ath11k_dp_tx_complete_msdu(s
 	    !(flags & ATH11K_SKB_F_NOACK_TID)) {
 	    	info->flags |= IEEE80211_TX_STAT_ACK;
 	    	bw_offset = arsta->last_tx_pkt_bw * ATH11K_TX_PKTS_BW_OFFSET;
-	    	info->status.ack_signal = ts.ack_rssi + ar->chan_noise_floor + bw_offset;
+	    	info->status.ack_signal = ar->rssi_offsets.rssi_offset ?
+					  ts.ack_rssi + ar->rssi_offsets.rssi_offset :
+					  ts.ack_rssi + ar->chan_noise_floor + bw_offset;
 	    	info->status.flags |= IEEE80211_TX_STATUS_ACK_SIGNAL_VALID;
 	}
 
--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -11744,9 +11744,15 @@ static void ath11k_mac_op_sta_statistics
 	}
 
 	if (arsta->rssi_comb) {
-		bw_offset = arsta->last_tx_pkt_bw * 3;
-		signal = arsta->rssi_comb + ar->chan_noise_floor + bw_offset;
-		sinfo->signal = signal;
+		if (ar->rssi_offsets.rssi_offset) {
+			sinfo->signal = arsta->rssi_comb +
+					ar->rssi_offsets.rssi_offset;
+		} else {
+			bw_offset = arsta->last_tx_pkt_bw * 3;
+			signal = arsta->rssi_comb + ar->chan_noise_floor +
+				 bw_offset;
+			sinfo->signal = signal;
+		}
 		sinfo->filled |= BIT_ULL(NL80211_STA_INFO_SIGNAL);
 	} else if (!signal &&
 	    arsta->arvif->vdev_type == WMI_VDEV_TYPE_STA &&
@@ -11756,7 +11762,13 @@ static void ath11k_mac_op_sta_statistics
 		signal = arsta->rssi_beacon;
 
 		if (signal) {
-			sinfo->signal = db2dbm ? signal : signal + ATH11K_DEFAULT_NOISE_FLOOR;
+			if (ar->rssi_offsets.rssi_offset) {
+				sinfo->signal = db2dbm ? signal :
+						signal + ar->rssi_offsets.rssi_offset;
+			} else {
+				sinfo->signal = db2dbm ? signal :
+						signal + ATH11K_DEFAULT_NOISE_FLOOR;
+			}
 			sinfo->filled |= BIT_ULL(NL80211_STA_INFO_SIGNAL);
 		}
 	}
@@ -11765,7 +11777,10 @@ static void ath11k_mac_op_sta_statistics
 		   "mac sta statistics db2dbm %u rssi comb %d rssi beacon %d\n",
 		   db2dbm, arsta->rssi_comb, arsta->rssi_beacon);
 
-	if (ar->chan_noise_floor) {
+	if (ar->rssi_offsets.rssi_offset) {
+		sinfo->signal_avg = ewma_avg_rssi_read(&arsta->avg_rssi) +
+				    ar->rssi_offsets.rssi_offset;
+	} else if (ar->chan_noise_floor) {
 		sinfo->signal_avg = ewma_avg_rssi_read(&arsta->avg_rssi) +
 				    ar->chan_noise_floor;
 	} else {
--- a/drivers/net/wireless/ath/ath11k/wmi.c
+++ b/drivers/net/wireless/ath/ath11k/wmi.c
@@ -8448,7 +8448,9 @@ static void ath11k_mgmt_rx_event(struct
 		status->freq = ieee80211_channel_to_frequency(rx_ev.channel,
 							      status->band);
 
-	status->signal = rx_ev.snr + ATH11K_DEFAULT_NOISE_FLOOR;
+	status->signal = ar->rssi_offsets.rssi_offset ?
+			 rx_ev.snr + ar->rssi_offsets.rssi_offset :
+			 rx_ev.snr + ATH11K_DEFAULT_NOISE_FLOOR;
 	status->rate_idx = ath11k_mac_bitrate_to_idx(sband, rx_ev.rate / 100);
 
 	hdr = (struct ieee80211_hdr *)skb->data;
